### ISSUE TITLE
Support HCL files

### ISSUE THREAD

**Issue by sixfeetover** [mozilla/sops#292]
Would love to see support for [HCL](https://github.com/hashicorp/hcl). HCL is used for a lot of configuration files as part of the Hashicorp stack (Vault, Terraform, etc).

**Comment by jvehent** [comment]
We don't use HCL at Mozilla, so it's unlikely we'll write support for it. However, we're happy to help review patches from the community.

**Comment by negz** [comment]
Note that all HCL can be represented as JSON (i.e. you can configure at least Terraform using JSON rather than HCL). We use `sops` to decrypt JSON files and pass those to Terraform, i.e.:
```bash
terraform apply -var-file=<(sops -d secret.tfvars.json)
```

There are also some HCL to JSON (and vice versa) converters around that might help workaround this use case.

**Comment by carlpett** [comment]
@sixfeetover @negz As another Terraform solution, I just created https://github.com/carlpett/terraform-sops, which lets you use a sops-encrypted file as a data source

**Comment by sixfeetover** [comment]
@carlpett Oh, very clever 👍

**Comment by jvehent** [comment]
This is awesome! @autrilla: could we host this under `sops/terraform` and add a use-case examples in the `examples` folder?

**Comment by autrilla** [comment]
@jvehent yeah, that'd be great, as long as @carlpett is OK with having their work go in our repo.

**Comment by starkers** [comment]
I'd really like to use sops for `.tfvars` files specifically..

Although yes its perfectly possible to implement a fancy `hcl2json` > `sops` process its just ugly and because of json we loose the ability to put comments in etc..

**Comment by carlpett** [comment]
@autrilla I'd be perfectly fine with that, but I think a better goal would be to have it hosted in the "official third-party" organization [terraform-providers](https://github.com/terraform-providers). That way, it'll be possible to download automatically, rather than doing a manual install.

@starkers I actually started off by looking at that, but found this approach seemed easier to use to me. But maybe I missed something! How was your intended workflow?

**Comment by negz** [comment]
@starkers Take a look at https://github.com/mozilla/sops/issues/292#issuecomment-360985371 - Terraform will happily load JSON encoded tfvars files.

From https://github.com/hashicorp/hcl:
> The API for HCL allows JSON as an input so that it is also machine-friendly (machines can generate JSON instead of trying to generate HCL).

**Comment by autrilla** [comment]
@carlpett yeah, that sounds better. Do you know what the process looks like to get it included there?

**Comment by so0k** [comment]
@carlpett thanks for sharing, as we were already using mozilla/sops - your provider allows our infra team to easily git clone and terraform apply without having to worry about custom make rules or tfvars

I did have to do the following to get this working on OSX:

```
go get github.com/carlpett/terraform-sops
mv $GOPATH/bin/terraform-sops $HOME/.terraform.d/plugins/terraform-provider-sops
```

**Comment by abeluck** [comment]
@carlpett  any progress on getting terraform-sops hosted with terraform-providers? What needs to be done to make that happen?

**Comment by rbrondgeest** [comment]
> I see now though that it is possible to sign up via a web form, so I'll get on it :)

Any progress on this @carlpett?

**Comment by binlab** [comment]
It will be nice to have

**Comment by abeluck** [comment]
@robertwe 

sops has no builtin support for HCL files.

If you're using terraform you can use @carlpett's [terraform-provider-sops](https://github.com/carlpett/terraform-provider-sops) to load sops files (json or yaml) into terraform. It works very well.

Also shout out to [terragrunt](https://terragrunt.gruntwork.io/docs/reference/built-in-functions/#sops_decrypt_file) which lets you decrypt sops files as part of your terraform+terragrunt workflow.

**Comment by robertwe** [comment]
@abeluck thx for info. I'm aware of that. I also considered this option. 
There is also a possibility to load `yaml` file directly in terraform.
The deal is I was searching for some super simple solution.

**Comment by sbaildon** [comment]
> > Note that all HCL can be represented as JSON (i.e. you can configure at least Terraform using JSON rather than HCL). We use `sops` to decrypt JSON files and pass those to Terraform, i.e.:
> > ```shell
> > terraform apply -var-file=<(sops -d secret.tfvars.json)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There are also some HCL to JSON (and vice versa) converters around that might help workaround this use case.
> 
> Hi @negz, I know your answer is 6y old, however I wondered if you can help with an issue I am facing. I had the same idea to use process substitution. We ran into a problem with the way how terraform interpretes the input data. [If the filename does not have a .json suffix, it will be interpreted as a hcl file](https://github.com/hashicorp/terraform/blob/33c0ae7f0947fc7fe83a35207591b661964bb9b1/internal/command/meta_vars.go#L230), so it was not possible to pass a file which gets interpreted as json via process substitution. Did you manage to solve this issue? I could not find a way to tell Terraform how to interprete the file besides the filename.

If using [fish](https://fishshell.com), one can  pipe into [psub](https://fishshell.com/docs/current/cmds/psub.html)

`terraform plan -var-file=(sops --decrypt main.tfvars.json | psub --suffix=.json)`
